Reorganize gateway general settings

pgalic96 opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #2807 but I think the main focus of this PR are gateway tests

#### Changes
<!-- What are the changes made in this pull request? -->

- Add `Collapse` wrapper around sections, reorganize sections


#### Testing

<!-- How did you verify that this change works? -->

- Manual, and cypress

##### Regressions

<!-- Please indicate features that this change could affect and how that was tested. -->

- Maybe some issues related to moving messages and validation schema to separate files


#### Checklist
<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

kschiffer commented:
Please always add a screenshot to PRs with visual changes.

bafonins commented on cypress/integration/console/gateways/general-settings.spec.js:
```diff
@@ -0,0 +1,123 @@
+
+  it('displays gateway values that are set', () => {
+    cy.loginConsole({ user_id: user.ids.user_id, password: user.password })
+    cy.createGateway(gateway, user.ids.user_id)
```
You can move it to `before`

bafonins commented on cypress/integration/console/gateways/general-settings.spec.js:
```diff
@@ -0,0 +1,123 @@
+      .should('not.be.disabled')
+      .and('have.attr', 'value')
+      .and('eq', gateway.ids.eui)
+    cy.findByLabelText('Gateway Name')
```
Please rename title for this field to `Gateway name` to be consistent with other forms.

bafonins commented on cypress/support/commands.js:
```diff
@@ -171,6 +171,20 @@ Cypress.Commands.add('augmentStackConfig', fns => {
   })
 })
 
+Cypress.Commands.add('createGateway', (gateway, userId) => {
```
Move this to other custom commands that make api calls.

bafonins commented on cypress/integration/console/gateways/general-settings.spec.js:
```diff
@@ -0,0 +1,123 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
```
Rename this file to `edit.spec.js`

bafonins commented on cypress/integration/console/gateways/general-settings.spec.js:
```diff
@@ -0,0 +1,123 @@
+    }
+  })
+
+  it('displays gateway values that are set', () => {
```
I would prefer checking values unset fields here as well. In this case we can assert on any default/empty values.

bafonins commented on cypress/integration/console/gateways/general-settings.spec.js:
```diff
@@ -0,0 +1,123 @@
+      `${Cypress.config('consoleRootPath')}/gateways/${gateway.ids.gateway_id}/general-settings`,
+    )
+    cy.findByText('General settings', { selector: 'h1' }).should('be.visible')
+    cy.findByText('Basic settings', { selector: 'h3' }).should('be.visible')
```
I would prefer using `cy.findByRole('heading', {name: 'Basic settings'})` here. We dont really care if it `h3`.

With `h1` tags it is different. We want to make sure that the page has at least one first level heading.

bafonins reviewed (changes requested):
I dont think we should change the creation form.

Please take a closer look at the end device general settings. There are some differences:

1. Delete button should be placed within the first section.
2. The user should be able to submit each section separately.
3. Remove the `General settings` heading. You can keep in the DOM for accessibility but hidden. See [visually-hidden mixin](https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.10/pkg/webui/styles/mixins.styl#L232-L242).
4. Remove `General settings` and `Gateway updates` headings.

kschiffer commented:
> I dont think we should change the creation form.

Indeed. We only use the collapsible sections in general settings form. If anything, we'd use a wizard like in the end device creation. Please revert these changes and use only a separation via `<Form.SubTitle title={m.mqttConfig} />`.

kschiffer commented on cypress/integration/console/gateways/general-settings.spec.js:
```diff
@@ -0,0 +1,123 @@
+    cy.findByLabelText('Gateway description')
+      .should('be.visible')
+      .and('have.text', gateway.description)
+    cy.findDescriptionByLabelText('Gateway description')
+      .should(
+        'contain',
+        'Optional gateway description; can also be used to save notes about the gateway',
+      )
+      .and('be.visible')
+    cy.findByLabelText('Gateway Server address')
+      .should('be.visible')
+      .and('have.attr', 'value', gateway.gateway_server_address)
+    cy.findDescriptionByLabelText('Gateway Server address')
+      .should('contain', 'The address of the Gateway Server to connect to')
+      .and('be.visible')
+    cy.findDescriptionByLabelText('Automatic updates')
+      .should('contain', 'Gateway can be updated automatically')
+      .and('be.visible')
```
I think checks like these are overly specific. We shouldn't care about the exact contents of the field descriptions. If anything, we can check whether they are visible, but even that, I wouldn't see it as a crucial matter for testing.

kschiffer commented on cypress/integration/console/gateways/general-settings.spec.js:
```diff
@@ -0,0 +1,123 @@
+      .should('contain', 'Channel for gateway automatic updates')
+      .and('be.visible')
+    cy.findByText('LoRaWAN options', { selector: 'h3' }).should('be.visible')
+    cy.findByText('LoRaWAN network-layer settings', { selector: 'p' }).should('be.visible')
```
Also here, try to avoid selecting by HTML element. These are implementation details that we don't want to depend on.

kschiffer commented on cypress/integration/console/gateways/general-settings.spec.js:
```diff
@@ -0,0 +1,123 @@
+      .should('be.visible')
+      .within(() => {
+        cy.findByText('Delete gateway', { selector: 'h1' }).should('be.visible')
+        cy.findByText(
+          `Are you sure you want to delete "${gateway.name}"? This action cannot be undone and it will not be possible to reuse the gateway ID.`,
+        ).should('be.visible')
```
Avoid checking by exact text. I don't want to update tests every time we change wording.

kschiffer commented on cypress/integration/smoke/gateways/delete.js:
```diff
@@ -0,0 +1,47 @@
+  cy.findByTestId('full-error-view').should('not.exist')
+
+  cy.location('pathname').should('eq', `${Cypress.config('consoleRootPath')}/gateways`)
+  cy.get('tbody').within(() => {
+    cy.findByRole('cell', { name: gateway.ids.gateway_id }).should('not.exist')
+  })
```
You might want to check whether the gateway was visible in the table before deleting it as well.

pgalic96 replied:
> See my other commend regarding the creation form.
> Regarding the tests: make sure that the checks stay closely scoped to the test-case. We shouldn't test every last pixel simply because we can. We need to ensure that tests do not lock us into one specific implementation more than necessary.
> It's not always trivial to do it and it shouldn't block you too much from building test cases, but please keep it in mind when writing tests.

I agree completely, but I am really not sure what to test for then. I think testing these forms this thoroughly (which also concerns the test you wrote for create application view) should be scoped to unit tests we already have, and then the tests should indeed be pixel by pixel, and changed once something is updated in the text itself.

That is why firstly I checked for all possible fields in the general settings form, then since it got quite long, I decided to remove the ones that are unset to keep it scoped to just checking if the values I set are reflected in the edit form that I am checking. But Bogdans (https://github.com/TheThingsNetwork/lorawan-stack/pull/3364#discussion_r507157778) also raised a concern about checking other fields as well. 

I like the way we use smoke tests, which exactly cover the use cases with concrete actions and a few UI checks to be sure the thing we test for really happened, but regarding the view tests, I am not so sure what are the tests about (I read the documentation in `CONTRIBUTING.md` and took a look at the already written tests) . cc @bafonins

kschiffer replied:
Addressing the code comments I've made on the tests is completely sufficient here. I'm just saying that e.g. checking on the exact text content of a description is too much of an irrelevant detail, while checking for unset values makes sense since it is relevant to the test-case (of checking correct form behavior).

I agree that the exact separation between our smoke and component tests are not straightforward and tend to overlap.

kschiffer replied:
Regarding the original comment here. I think since this is a smoke test, it's fine to use a wider scope of checks that resemble more of a complete user story. Adding another check here to verify that the gateway was there in the first place is a cheap way to improve the expressiveness of this test.

kschiffer commented on pkg/webui/console/components/gateway-data-form/index.js:
```diff
@@ -190,120 +121,133 @@ class GatewayDataForm extends React.Component {
+            component={KeyValueMap}
+            description={sharedMessages.attributeDescription}
+          />
+          <Message component="h4" content={sharedMessages.gatewayUpdateOptions} />
+          <Form.Field
+            title={sharedMessages.automaticUpdates}
+            name="auto_update"
+            component={Checkbox}
+            description={m.autoUpdateDescription}
+          />
+          <Form.Field
+            title={sharedMessages.channel}
+            description={m.updateChannelDescription}
+            placeholder={sharedMessages.stable}
+            name="update_channel"
+            component={Input}
+            autoComplete="on"
+          />
```
I feel like these could either be put into a separate section or be merged with `LoRaWAN Options`. @benolayinka do you have an idea for a better separation?

benolayinka replied:
I don't think it matters. They aren't really LoRaWAN options. What kind of updates are these, though? And the "stable" prompt can be improved

<img width="685" alt="Screenshot 2020-10-20 at 16 46 13" src="https://user-images.githubusercontent.com/6963436/96603135-062f1080-12f4-11eb-9b1c-62e078120127.png">

kschiffer reviewed (changes requested):
See my other commend regarding the creation form.
Regarding the tests: make sure that the checks stay closely scoped to the test-case. We shouldn't test every last pixel simply because we can. We need to ensure that tests do not lock us into one specific implementation more than necessary. 
It's not always trivial to do it and it shouldn't block you too much from building test cases, but please keep it in mind when writing tests.

KrishnaIyer commented:
Are there some screenshots? It would be nice to know what's changing (since I'm mostly responsible for gateways).

pgalic96 commented:
> Are there some screenshots? It would be nice to know what's changing (since I'm mostly responsible for gateways).

It is a minor reorganization of attributes in 2 sections, subject to changes:

![Screenshot 2020-10-20 at 6 20 38 PM](https://user-images.githubusercontent.com/19795814/96615084-21544d00-1301-11eb-84e5-448d66409194.png)

![Screenshot 2020-10-20 at 6 21 01 PM](https://user-images.githubusercontent.com/19795814/96615120-2fa26900-1301-11eb-9777-88ad599bd166.png)

pgalic96 committed:
console: Move collapse to components

pgalic96 committed:
console: Reorganize gateway general settings

pgalic96 committed:
console: Add gateway cypress tests

pgalic96 committed:
console: Adapt gateway-data-form

pgalic96 commented:
@kschiffer @bafonins I believe your comments are all addressed now. Also regarding tests, I tried focusing on only the relevant fields and tried to minimize checking irrelevant label values and such.

bafonins reviewed (commented):
```
Uncaught ReferenceError: m is not defined
    at GatewayDataForm.render (index.js?21dd:224)
    at finishClassComponent (react-dom.development.js?7f13:17036)
    at updateClassComponent (react-dom.development.js?7f13:16991)
    at beginWork$1 (react-dom.development.js?7f13:18502)
    at HTMLUnknownElement.callCallback (react-dom.development.js?7f13:348)
    at Object.invokeGuardedCallbackDev (react-dom.development.js?7f13:398)
    at invokeGuardedCallback (react-dom.development.js?7f13:455)
    at beginWork$$1 (react-dom.development.js?7f13:23214)
    at performUnitOfWork (react-dom.development.js?7f13:22205)
    at workLoopSync (react-dom.development.js?7f13:22182)
```

bafonins commented on pkg/webui/console/views/gateway-general-settings/gateway-general-settings.styl:
```diff
@@ -0,0 +1,16 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+.visually-hidden
```
```suggestion
.title
```

bafonins replied:
```suggestion
.title-hidden
```

bafonins replied:
```suggestion
.page-title
```

bafonins replied:
```suggestion
.page-title-hidden
```

bafonins commented on cypress/integration/smoke/gateways/delete.js:
```diff
@@ -0,0 +1,51 @@
+  }
+  cy.loginConsole({ user_id: user.ids.user_id, password: user.password })
+  cy.createGateway(gateway, user.ids.user_id)
+  cy.visit(`${Cypress.config('consoleRootPath')}/gateways`)
```
We want to test that the user is able to reach this page via UI

bafonins commented on cypress/integration/smoke/gateways/delete.js:
```diff
@@ -0,0 +1,51 @@
+  cy.get('tbody').within(() => {
+    cy.findByRole('cell', { name: gateway.ids.gateway_id }).should('be.visible')
+  })
+  cy.visit(
```
We want to test that the user is able to reach this page via UI

bafonins commented on pkg/webui/components/key-value-map/entry.js:
```diff
@@ -84,6 +84,7 @@ class Entry extends React.Component {
     return (
       <div className={style.entriesRow}>
         <Input
+          data-test-id={'key'}
```
Why not relying on the `name` attribute which is unique? With this approach each entry will have the same `data-test-id` attribute.

bafonins commented on pkg/webui/components/key-value-map/entry.js:
```diff
@@ -94,6 +95,7 @@ class Entry extends React.Component {
           code
         />
         <Input
+          data-test-id={'value'}
```
Why not relying on the `name` attribute which is unique? With this approach each entry will have the same `data-test-id` attribute.

bafonins reviewed (changes requested):
Im getting the following error when submitting general settings form without any any changes:
```
{
  "code": 3,
  "message": "error:pkg/errors:validation (invalid `gateway`: embedded message failed validation)",
  "details": [
    {
      "@type": "type.googleapis.com/ttn.lorawan.v3.ErrorDetails",
      "namespace": "pkg/errors",
      "name": "validation",
      "message_format": "invalid `{field}`: {reason}",
      "attributes": {
        "field": "gateway",
        "name": "UpdateGatewayRequestValidationError",
        "reason": "embedded message failed validation"
      },
      "correlation_id": "7c6fc862056b4fb2af12d27b87439d45",
      "cause": {
        "namespace": "pkg/errors",
        "name": "validation",
        "message_format": "invalid `{field}`: {reason}",
        "attributes": {
          "field": "version_ids",
          "name": "GatewayValidationError",
          "reason": "embedded message failed validation"
        },
        "cause": {
          "namespace": "pkg/errors",
          "name": "validation",
          "message_format": "invalid `{field}`: {reason}",
          "attributes": {
            "field": "model_id",
            "name": "GatewayVersionIdentifiersValidationError",
            "reason": "value does not match regex pattern \"^[a-z0-9](?:[-]?[a-z0-9]){2,}$\""
          },
          "code": 3
        },
        "code": 3
      },
      "code": 3
    }
  ]
}
```

kschiffer commented on pkg/webui/console/views/gateway-general-settings/basic-settings-form/index.js:
```diff
@@ -0,0 +1,200 @@
+    mayDeleteGateway,
+  } = props
+
+  const [error, setError] = React.useState('')
```
```suggestion
  const [error, setError] = React.useState(undefiend)
```

Nit, but it rellay doesn't make sense that we always use an empty string for this.

kschiffer commented on pkg/webui/console/views/gateway-general-settings/basic-settings-form/index.js:
```diff
@@ -0,0 +1,200 @@
+      const castedValues = validationSchema.cast(values)
+      const updatedValues = diff(initialValues, castedValues)
+
+      setError('')
```
```suggestion
      setError(undefined)
```

kschiffer commented on pkg/webui/console/views/gateway-general-settings/index.js:
```diff
@@ -90,131 +81,100 @@ export default class GatewayGeneralSettings extends React.Component {
     return (
       <Container>
-        <PageTitle title={sharedMessages.generalSettings} />
+        <PageTitle className={style.visuallyHidden} title={sharedMessages.generalSettings} />
```
Theres a `hideHeading` prop to do just that.

```suggestion
        <PageTitle title={sharedMessages.generalSettings} hideHeading />
```

kschiffer commented on pkg/webui/console/views/gateway-general-settings/gateway-general-settings.styl:
```diff
@@ -0,0 +1,16 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
```
File can be removed. See my other comment on the `hideHeading` prop.

kschiffer reviewed (changes requested):
Not much to add. Please fix the submit error.

pgalic96 commented:
@bafonins @kschiffer I fixed the submit error, however right now the behavior where `attributes` get appended on every `Submit` (which is also what is happening right now in live) is not ideal. How I implemented it before was ideal, but there seems to be some validation bug on the backend that doesn't allow for pushing empty payloads like with application form for example. I submitted an issue https://github.com/TheThingsNetwork/lorawan-stack/issues/3406 and once it is solved I will reimplement the previous solution.

bafonins commented on cypress/integration/console/gateways/edit.spec.js:
```diff
@@ -0,0 +1,119 @@
+    cy.findByLabelText('Gateway description')
+      .should('be.visible')
+      .and('have.text', gateway.description)
+    cy.findDescriptionByLabelText('Gateway description').should('be.visible')
```
```suggestion
    cy.findDescriptionByLabelText('Gateway description')
      .should(
        'contain',
        'Optional gateway description; can also be used to save notes about the gateway',
      )
      .and('be.visible')
```

bafonins commented on cypress/integration/console/gateways/edit.spec.js:
```diff
@@ -0,0 +1,119 @@
+    cy.findByLabelText('Gateway Server address')
+      .should('be.visible')
+      .and('have.attr', 'value', gateway.gateway_server_address)
+    cy.findDescriptionByLabelText('Gateway Server address').should('be.visible')
```
```suggestion
    cy.findDescriptionByLabelText('Gateway Server address')
      .should('contain', 'The address of the Gateway Server to connect to').and('be.visible')
```

bafonins commented on cypress/integration/console/gateways/edit.spec.js:
```diff
@@ -0,0 +1,119 @@
+    cy.findByLabelText('Gateway status')
+      .should('be.visible')
+      .and('have.attr', 'value', 'false')
+    cy.findDescriptionByLabelText('Gateway status').should('be.visible')
```
```suggestion
    cy.findDescriptionByLabelText('Gateway status')
      .should('contain', 'The status of this gateway may be publicly displayed')
      .and('be.visible')
```

bafonins commented on cypress/integration/console/gateways/edit.spec.js:
```diff
@@ -0,0 +1,119 @@
+    cy.findByLabelText('Automatic updates')
+      .should('be.visible')
+      .and('have.attr', 'value', 'false')
+    cy.findDescriptionByLabelText('Automatic updates').should('be.visible')
```
```suggestion
    cy.findDescriptionByLabelText('Automatic updates')
      .should('contain', 'Gateway can be updated automatically')
      .and('be.visible')
```

bafonins commented on cypress/integration/console/gateways/edit.spec.js:
```diff
@@ -0,0 +1,119 @@
+      .should('be.visible')
+      .and('have.attr', 'placeholder')
+      .and('eq', 'Stable')
+    cy.findDescriptionByLabelText('Channel').should('be.visible')
```
```suggestion
    cy.findDescriptionByLabelText('Channel')
      .should('contain', 'Channel for gateway automatic updates')
      .and('be.visible')
```

bafonins commented on pkg/webui/components/unit-input/index.js:
```diff
@@ -98,6 +98,7 @@ class UnitInput extends React.PureComponent {
       <React.Fragment>
         <div className={classnames(className, style.container)}>
           <Input
+            data-test-id="unit-input"
```
If we have several of `UnitInput` component on the same view? `cy.findByTestId` will return all of them.

bafonins commented on pkg/webui/components/page-title/index.js:
```diff
@@ -30,7 +30,7 @@ const PageTitle = ({ title, values, tall, className, hideHeading, children }) =>
     pageTitle
   ) : (
-    <Row>
+    <Row className={className}>
```
Do you still need this? Cannot find any usage of `PageTitle` with `className` prop

bafonins reviewed (commented):
Works. Please adjust the tests and we can merge this

pgalic96 committed:
dev: Add messages

pgalic96 commented:
should all be good now.
